api.lua

```diff
@@ -210,28 +233,23 @@ local function fire(stack, player)
 			end)
 		end
 
-		minetest.after(time, function(obj, pos, look_dir, gun_def)
-			local pointed = get_pointed_thing(pos, look_dir, gun_def)
-			if pointed and pointed.type == "object" then
-				local target = pointed.ref
-				if target:get_player_name() ~= obj:get_player_name() then
-					local point = pointed.intersection_point
-					local dmg = config.base_dmg * gun_def.dmg_mult
-
-					-- Add 50% damage if headshot
-					if point.y > target:get_pos().y + 1.2 then
-						dmg = dmg * 1.5
-					end
-
-					-- Add 20% more damage if player using scope
-					if gunslinger.__scopes[obj:get_player_name()] then
-						dmg = dmg * 1.2
-					end
-
-					target:punch(obj, nil, {damage_groups = {fleshy = dmg}})
-				end
-			end
-		end, player, pos1, dir, def)
+		-- handle hitscan / projectile weapons
+		if def.hit_type == "hitscan" then
+			hit_target(player, pos1, dir, def, initial_pthing)
+		else
+			--[[
+				Perform "deferred raycasting" to mimic projectile entities, without
+				actually using entities:
+					- Perform initial raycast to get position of target if it exists
+					- Calculate time taken for projectile to travel from gun to target
+					- Perform actual raycast after the calculated time
+
+				This process throws in a couple more calculations and an extra raycast,
+				but the vastly improved realism at the cost of a negligible performance
+				hit is always great to have.
+			]]
+			minetest.after(time, hit_target, player, pos1, dir, def, 0)
+		end
```

Comment:
PR looks good overall. The way it integrates into the existing API is nice.

One thing I'd suggest is to isolate various firing mechanisms more, and preferably move them into their separate methods. This would result in more readable code than implementing entire firing systems within `if-else` blocks.

i.e.

```lua
local function fire(...)

.
.

if def.hit_type == "" then
    hitscan(...)
else if def.hit_type == "deferred_raycasting"
    ...
end

.
.

end
```

Don't bother moving the Deferred raycasting code to a separate function though - it'll be removed very soon. :)

Reply:
roger that